### PR TITLE
service/header: Header Exchange implemented over Core Node

### PR DESCRIPTION
**Pull request by renaynay**
This PR will allow header service to sync headers from the core node in order to sync up to the chain. This is necessary for bootstrap peers.

### CODE REVIEW

File: service/header/core_exchange.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+func (ce *CoreExchange) RequestHeader(ctx context.Context, height uint64) (*ExtendedHeader, error) {
+	intHeight := int64(height) + 1
+	block, err := ce.fetcher.GetBlock(ctx, &intHeight)
+	if err != nil {
+		return nil, err
+	}
+	return ce.generateExtendedHeaderFromBlock(block)
+}
+
+func (ce *CoreExchange) RequestHeaders(ctx context.Context, origin, amount uint64) ([]*ExtendedHeader, error) {
```
Comment (Wondertan) [inline service/header/core_exchange.go:40]: This will be so slooow, but it is fine for now. Just for reference, we should use BlockSearch API instead

**Review by Wondertan — APPROVED** [review]

**Review by liamsi — APPROVED** [review]

File: service/header/core_exchange.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+func NewCoreExchange(fetcher *core.BlockFetcher, dag format.DAGService) *CoreExchange {
+	return &CoreExchange{
+		fetcher:    fetcher,
+		shareStore: dag,
+	}
+}
+
+func (ce *CoreExchange) RequestHeader(ctx context.Context, height uint64) (*ExtendedHeader, error) {
+	intHeight := int64(height) + 1
```
Comment (adlerjohn) [inline service/header/core_exchange.go:32]: Are we sure that passing `math.MaxUint64` won't result in some weird behavior in the node?

Comment (liamsi) [reply service/header/core_exchange.go:32]: MaxInt64 would already overflow here. Maybe just add a TODO to make sure that conversion is safe.

Comment (adlerjohn) [reply service/header/core_exchange.go:32]: Yeah a TODO should suffice. If we're not careful, someone will be able to crash a node by simply asking for a block at a large height, for example, which is pretty bad.

Comment (Wondertan) [reply service/header/core_exchange.go:32]: That is like saying: Are we sure that passing `nil` to `reader.Read` won't result in some weird behavior in the node?

I mean this is like an unreal case that should be possible until we reach such height.

Comment (adlerjohn) [reply service/header/core_exchange.go:32]: In this particular case, requesting headers is done by an untrusted peer! So you probably don't want that to be able to _crash_ your node. If it were something local and trusted, then it'd be a non-issue. Or is this function only callable internally?

Comment (liamsi) [reply service/header/core_exchange.go:32]: It is talking to a trusted core node indeed.

Comment (Wondertan) [reply service/header/core_exchange.go:32]: The reason is that we count blocks from `0`, while tendermint from 1

Comment (Wondertan) [reply service/header/core_exchange.go:32]: Btw, +1 there is because we count blocks from 0 while Tendermint from 1. What should we stick to?

Comment (Wondertan) [reply service/header/core_exchange.go:32]: > In this particular case, requesting headers is done by an untrusted peer! So you probably don't want that to be able to crash your node. If it were something local and trusted, then it'd be a non-issue. Or is this function only callable internally?

Ah, I got your point now. Well, the CoreExchange is not going to be part of a publicly facing API. Though, there might be a case where this happens.

**Pull request merged**